### DISCUSSION TITLE
compat.Pytorch2River

### DISCUSSION THREAD

**Discussion by batman7090** [online-ml/river#972] [Q&A] [answered]
hello all,
I am getting a TypeError: forward() got an unexpected keyword argument 'n_features'  I am not getting how to pass this parameters into the build model function could anybody shed some light on this? I have attached the code below.

Thank you in advance.

```py
from river import compat
from river import datasets
from river import evaluate
from river import metrics
from river import preprocessing
from torch import nn
from torch import optim
import torch
import pandas as pd
from river import stream
import itertools

_ = torch.manual_seed(0)
n_features=3

def build_torch_mlp_regressor(n_features):
  net = nn.Sequential(
      nn.Linear(n_features,45),
      nn.Linear(45,50),
      nn.Linear(50,50),
      nn.Linear(50,3)
  )
  return net
model = compat.PyTorch2RiverRegressor(build_fn=build_torch_mlp_regressor(n_features),
                                      loss_fn=nn.MSELoss,
                                      optimizer_fn=optim.Adam)

df = pd.read_csv('/content/drive/MyDrive/data/Final_data/one_lakh_more_accurate.csv')
train_size = 100


for xb in pd.read_csv('/content/drive/MyDrive/data/Final_data/soft_robotics_2_scaled.csv', chunksize=50):
  y = xb.loc[:,'X':'Z']
  x = xb.loc[:,'P_set_1':'P_set_3']
  model.predict_one(x)
  model.learn_one(x,y)
```
Error I got

T

> ypeError                                 Traceback (most recent call last)
> [<ipython-input-38-30a190da9b78>](https://localhost:8080/#) in <module>()
>      10   y = xb.loc[:,'X':'Z']
>      11   x = xb.loc[:,'P_set_1':'P_set_3']
> ---> 12   model.predict_one(x)
>      13   model.learn_one(x,y)
>      14 
> 
> 2 frames
> [/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
>    1108         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
>    1109                 or _global_forward_hooks or _global_forward_pre_hooks):
> -> 1110             return forward_call(*input, **kwargs)
>    1111         # Do not call functions when jit is used
>    1112         full_backward_hooks, non_full_backward_hooks = [], []
> 
> TypeError: forward() got an unexpected keyword argument 'n_features'

**Comment by MaxHalford** [comment]
Basically `build_fn` expects to be passed a function, and you're actually calling the function. This should work:

```py
model = compat.PyTorch2RiverRegressor(build_fn=build_torch_mlp_regressor
                                      loss_fn=nn.MSELoss,
                                      optimizer_fn=optim.Adam)
```

Why is this? Because in online learning you the user should never have to worry about the number of features in the data stream. Models should handle this for you.

Note that we'll be removing PyTorch compatibility in the next release. It will instead be located in [river-extra](https://github.com/online-ml/river-extra).

**Reply by batman7090** [reply]
@MaxHalford no it is not working could you please share with me the example of how can iter over CSV file with train test loop?
can I use PyTorch2RiverRegressor for multi-target regression? 

thank you in advance

**Comment by MaxHalford** [comment, marked as answer]
> could you please share with me the example of how can iter over CSV file with train test loop?

Take a look at [`iter_csv`](https://riverml.xyz/0.11.1/api/stream/iter-csv/) as well as [`iter_pandas`](https://riverml.xyz/0.11.1/api/stream/iter-pandas/).

> can I use PyTorch2RiverRegressor for multi-target regression?

No, it doesn't support that. But I encourage you to consider the [`multioutput` module](https://riverml.xyz/0.11.1/api/overview/#multioutput_1). It contains model wrappers, which you can wrap around any regression model, which includes `PyTorch2RiverRegressor`.

**Reply by batman7090** [reply]
@MaxHalford,  thank you so much it is working now.
